src/encoded/tests/test_search.py

```diff
@@ -543,22 +555,83 @@ def test_collection_actions_filtered_by_permission(workbook, testapp, anontestap
+        # This might get a 404 if not enough time has elapsed, so try a few times before giving up.
+        #
+        # We retry a lot of times because it's still fast if things are working quickly, but if it's
+        # slow it's better to wait than fail the test. Slowness is not what we're trying to check for here.
+        # And even if it's slow for one item, that same wait time will help others have time to catch up,
+        # so it shouldn't be slow for others. At least that's the theory. -kmp 27-Jan-2021
+        extra = "&status=deleted" if deleted else ""
+        return client.get('/%s?limit=all%s' % (item_type, extra), status=[200, 301]).follow()
+
+
+    CONSIDER_DELETED = True
+    DELETED_SEEN = {}
+
+    @classmethod
+    def reset_deleted(cls):
+        cls.DELETED_SEEN = {}
+
+    @classmethod
+    def deleted_seen(cls):
+        total_deleted = 0
+        for item_type, item_deleted_count in cls.DELETED_SEEN.items():
+            ignored(item_type)
+            total_deleted += item_deleted_count
+        return total_deleted
+
+    @classmethod
+    def get_all_items_of_type(cls, client, item_type):
+        if cls.CONSIDER_DELETED:
+            try:
+                res = cls.check_item_type(client, item_type)
+                items_not_deleted = res.json.get('@graph', [])
+            except webtest.AppError:
+                items_not_deleted = []
+            try:
+                res = cls.check_item_type(client, item_type, deleted=True)
+                items_deleted = res.json.get('@graph', [])
+            except webtest.AppError:
+                items_deleted = []
+            if items_deleted:
+                cls.DELETED_SEEN[item_type] = items_deleted
+            else:
+                cls.DELETED_SEEN.pop(item_type, None)  # delete entry if present but we want it empty
+            return items_not_deleted + items_deleted
+        else:
+            res = cls.check_item_type(client, item_type)
+            items_not_deleted = res.json.get('@graph', [])
+            return items_not_deleted
 
 
 @pytest.mark.flaky
 def test_index_data_workbook(app, workbook, testapp, indexer_testapp, htmltestapp):
     es = app.registry['elasticsearch']
     # we need to reindex the collections to make sure numbers are correct
     create_mapping.run(app, sync_index=True)
-    # check counts and ensure they're equal
-    testapp_counts = testapp.get('/counts')
-    split_counts = testapp_counts.json['db_es_total'].split()
-    assert(int(split_counts[1]) == int(split_counts[3]))  # 2nd is db, 4th is es
+    retried = False
+    while True:
+        # check counts and ensure they're equal
+        testapp_counts = testapp.get('/counts')
+        # e.g., {"db_es_total": "DB: 748 ES: 748 ", ...}
+        db_es_total = testapp_counts.json['db_es_total']
+        split_counts = db_es_total.split()
+        db_total = int(split_counts[1])
+        es_total = int(split_counts[3])
+        if db_total == es_total or retried:
```

Comment:
We used to just do an assert of this, but that made the test end before we got better data. It also meant that if the counts just hadn't stabilized, we'd lose falsely, so now we go ahead and loop once. That seems to be what fixes this to be a more stable test.